### ISSUE TITLE
Authorization

### ISSUE THREAD

**Issue by daviddykeuk** [turbosonic/api-gateway#15]
Need to add scopes and rules to endpoints via configuration

**Comment by daviddykeuk** [comment]
## 🔏 Scopes

Need to add scopes to endpoints so that consuming clients must ask the user for that scope before the client being able to call the endpoint, this needs to be done in a simple way, Auth0 can facilitate the scope claims in the JWT, we just need to process them, first thoughts on config are...

```yaml
endpoints:
  - url: /posts
    methods:
      - method: GET
        scopes:            # <<< new property
          - read:posts     # <<< new property   
          - create:posts   # <<< new property 
        destination:
          name: health
          host: http://localhost:3000
          url: /posts
```

Scopes is an array because multiple scopes may be required to access a single endpoint, particularly if they are bundled requests and going to multiple endpoints. << **I've changed my mind on this below**

**Comment by daviddykeuk** [comment]
## ⚠️ Safety first
No scope should mean deny everyone, you have to say that the endpoint is open to save yourself from accidents, the below would open up an endpoint to all
```yaml
scopes:
  - *
```

**Comment by daviddykeuk** [comment]
## 🔐 Roles
By enriching the claims on the `access_token` from Auth0 with roles, users roles can be administered by the identity provider and dealt with as trusted by the API.

e.g. roles can come in the access token as
```json
{
  "https://api.turbosonic.com/roles": [
    "admin",
    "developer"
  ]
}
```

n.b. roles always need a namespace and we can set this to the same as the audience used during authentication

Then we can use those roles to decide if the user can access an endpoint via configuration...
```yaml
endpoints:
  - url: /posts
    methods:
      - method: GET
        scopes:            
          - read:posts     
          - create:posts    
        roles:          # <<< new property
          - admin       # <<< new property
          - developer   # <<< new property
        destination:
          name: health
          host: http://localhost:3000
          url: /posts
```

There's an array of roles because the user needs to have at least one of them to access this endpoint, however, this clashes with the way of doing scopes, so I think it's best to do the same with scopes and allow access to the endpoint if ANY of those scopes are in the claims and limit bundled endpoints in another way and in another issue.

**Comment by daviddykeuk** [comment]
## ⚠️ Safety first
No role should mean deny everyone, you have to say that the endpoint is open to save yourself from accidents, the below would open up an endpoint to all
```yaml
roles:
  - *
```

**Comment by daviddykeuk** [comment]
closed by #16
